[red-knot] document test framework (#13695)
astral-sh/ruff: crates/red_knot_test/README.md

The simplest possible test suite consists of just a single test, with a single file.

````markdown
```py
reveal_type(1)  # revealed: Literal[1]
```
````

When running this test, the mdtest framework will write a file with these contents to the default
file path (`/src/test.py`) in its in-memory file system, run a type check on that file, and then
match the resulting diagnostics with the assertions in the test. Assertions are in the form of
Python comments. If all diagnostics and all assertions are matched, the test passes; otherwise, it
fails.

## Assertions

Two kinds of assertions are supported, `# revealed:` (shown above) and `# error:`.

### Assertion kinds

#### revealed

A `# revealed:` assertion should always be paired with a call to the `reveal_type` utility, which
reveals (via a diagnostic) the inferred type of its argument (which can be any expression). The text
after `# revealed:` must match exactly with the displayed form of the revealed type of that
expression.

The `reveal_type` function can be imported from the `typing` standard library module (or, for older
Python versions, from the `typing_extensions` pseudo-standard-library module -- it is a third-party

> module, but typeshed, and thus type checkers also, treat it as part of the standard library):

Author: Using a footnote for the digression is a good idea, I'll do that.

> I'd wager that 99% of the time you'll want to import it from `typing_extensions` rather than `typign`

I don't think this is true; I think rather that 99% of tests will not import it at all, and the only ones that do will be a few tests that are explicitly testing the handling of imported vs pseudo-builtin `reveal_type`, and of those few tests probably roughly half (i.e. just one or two tests each) will import it each way.

I think the more important consideration here is that the preferred and modern thing to do, if on a Python version where it's possible, is to import from `typing`, so that's what should be featured, and the legacy approach should be a parenthetical.

```py
from typing import reveal_type

reveal_type("foo")  # revealed: Literal["foo"]
```

For convenience, type checkers also pretend that `reveal_type` is a built-in, so that this import is
not required. Using `reveal_type` without importing it issues a diagnostic warning that it was used
without importing it, in addition to the diagnostic revealing the type of the expression.

The `# revealed:` assertion must always match a revealed-type diagnostic, and will also match the
undefined-reveal diagnostic, if present, so it's safe to use `reveal_type` in tests either with or
without importing it. (Style preference is to usually not import it in tests, unless specifically

> testing something about the behavior of importing it.)

Author: I think the current text accurately and precisely describes how they are handled, and this information is relevant to authors/readers of tests, because the details of how they are handled can be visible to authors of tests (for example, if you do `rt = return_type` and then `rt(foo) # revealed: ...` on another line, you will see unimported-reveal diagnostic on the `rt = return_type` line, and it won't be matched by the `# revealed: ...` assertion. And we will have tests that depend on these distinctions (the tests of the undefined-reveal diagnostic itself.)

The proposed new text isn't noticeably shorter, and it has the disadvantage of being wrong :) It's simply not true that "this diagnostic is never emitted" in tests using mdtest.

> testing something about the behavior of importing it.)

Author: Reading this again, I do think the text could be re-ordered a bit to emphasize the most important thing (`# revealed: ` matches a revealed type from `reveal_type` and then follow up with the stuff about undefined-reveal separately; I'll look at an edit along those lines.

#### error

A comment beginning with `# error:` is an assertion that a type checker diagnostic will
be emitted, with text span starting on that line. If the comment is simply `# error:`, this will

> match any diagnostic. The matching can be narrowed in three ways:

Author: I agree that we should probably never use this, it's just the most natural implementation. Personally I think it's fine to just avoid ever using such assertions, but allow them. If we want to prohibit them, we have to decide what the precise rule should be. Is an assertion that only asserts the column OK? Or should the error code always be required? Or should it be "either error code or message-contains" that is required?

> match any diagnostic. The matching can be narrowed in three ways:

Reviewer: > Personally I think it's OK to allow such assertions but just avoid using them.

I think it's inevitable that either one of us or a contributor adds one (either deliberately or accidentally) at some point, if we allow it!

> Or should it be "either error code or message-contains" that is required?

I would say this

> match any diagnostic. The matching can be narrowed in three ways:

Author: Ok, that seems reasonable, I'll look at making that behavior change in this PR.

- `# error: 8` requires that the matched diagnostic's text span begins on column 8 (one-indexed) of
    this line.
- `# error: [invalid-assignment]` requires that the matched diagnostic have the rule code
    `invalid-assignment`.

> - `# error: "Some text"` requires that the matched diagnostic's full message contain the text `Some text`.

Reviewer 2: Assuming that the quotes are required, can we clarify that part somewhere here?

Any combination of two or all three of these can be used, but they must go in order: first column,
if present; then rule code, if present, then contains-text, if present. For example, an assertion
using all three would look like `# error: 8 [invalid-assignment] "Some text"`.

Error assertions in tests intended to test type checker semantics should use primarily rule-code
assertions, with occasional contains-text assertions where needed to disambiguate.

Some tests may exist in order to assert on the correct text-range or text of a diagnostic; these
should use the appropriate assertions.

### Assertion locations

An assertion comment may be a line-trailing comment, in which case it applies to the line it is on:

```py
x: str = 1  # error: [invalid-assignment]
```

Or it may be a comment on its own line, in which case it applies to the next non-assertion-comment

> line:

Reviewer 2: I find "non-assertion-comment" a bit confusing because at first I thought it meant a non-assertion _comment_, so maybe the next statement / code line?

```py
# error: [invalid-assignment]
x: str = 1
```

Multiple assertions applying to the same line may be stacked:

```py
# error: [invalid-assignment]
# revealed: Literal[1]
x: str = 1
```

Intervening empty lines or non-assertion comments are not allowed; an assertion stack must be one
assertion per line, immediately following each other, with the line immediately following the last
assertion the line of source code emitting the matched diagnostics.

## Multi-file tests

Some tests require multiple files, with imports from one file to another. Multiple fenced code

> blocks represent multiple embedded files. Since files must have unique names, only zero or one files

Reviewer 2: Should it be "only one file can use..." instead of "only zero or one files can use..."?

can use the default name of `/src/test.py`. Other files must explicitly specify their file name:

````markdown
```py
from b import C

> reveal_type(C)  # Literal[C]

Reviewer 2: Did you mean?
```suggestion
reveal_type(C)  # revealed: Literal[C]
```

Author: I would love to do that! But probably not a priority :)

```

```py path=b.py
class C: pass
```
````

> Relative file names are always relative to the "workspace root", which is also an import root.

Author: I can try to make it more clear, maybe with reference to `sys.path`.

The default workspace root is `/src/`. Currently it is not possible to customize this in a test, but
this is a feature we will want to add in the future.

So the above test creates two files, `/src/test.py` and `/src/b.py`, and sets the workspace root to
`/src/`, allowing `test.py` to import from `b.py` using the module name `b`.

## Multi-test suites

> A single test suite (markdown file) can contain multiple tests, by demarcating them using Markdown

Author: Agreed, I prefer "Markdown" and will aim to be consistent with that.

header lines:

````markdown
# Same-file invalid assignment

```py
x: int = "foo"  # error: [invalid-assignment]
```

# Cross-file invalid assignment

```py
from b import y
x: int = y  # error: [invalid-assignment]
```

```py path=b.py
y = "foo"
```
````

This test suite contains two tests, one named "Same-file invalid assignment" and the other named
"Cross-file invalid assignment". The first test involves only a single embedded file, and the second
test involves two embedded files.

The tests are run independently, in independent in-memory file-systems and with new red-knot salsa
databases (so each is a from-scratch run of the type checker, with no data persisting from any
previous test.)

> Due to Rust test runner limitations, an entire test suite (Markdown file) is run as a single Rust

Author: I'll probably be more specific here and just say due to `cargo test` limitations

> test, so it's not possible to select individual tests within it to run.

Author: It's an inherent limitation of `cargo test` that there is no way to dynamically create tests. I think to directly address this limitation we'd have to either write our own test runner, or use a build script to code-generate tests. Neither of those are planned. 

What I think is more likely to indirectly address this problem is that we could add some kind of "focus" indicator that's easy to add and remove from a test file that will cause it to only run that test from that file. Or we could do something similar that works via env var.

## Documentation of tests

Arbitrary Markdown syntax (including of course normal prose paragraphs) is permitted (and ignored by
the test framework) between fenced code blocks. This allows naturally documenting test intentions:

````markdown
Assigning a string to a variable annotated as `int` is not permitted:

```py
x: int = "foo"  # error: [invalid-assignment]
```
````

## Planned features

There are some designed features that we intend for the test framework to have, but have not yet
implemented:

### Multi-line diagnostic assertions

We may want to be able to assert that a diagnostic spans multiple lines, and which columns it
begins/ends on. The planned syntax for this uses `<<<` and `>>>` to mark the start and end lines for
an assertion:

```py
(3  # error: 2 [unsupported-operands] <<<
  +
 "foo")  # error: 6 >>>
```

The column assertion `6` on the ending line should be optional.

In cases of overlapping such assertions, resolve ambiguity using more brackets: `<<<<` begins an
assertion ended by `>>>>`, etc.

### Non-Python files

Some tests may need to specify non-Python embedded files: typeshed `stdlib/VERSIONS`, `pth` files,
`py.typed` files, `pyvenv.cfg` files...

> We should allow specifying any of these using the `text` language in the code block tag string:

Author: I was trying to steer away from language that suggests an ironclad commitment to do any of these things precisely as described...

> We should allow specifying any of these using the `text` language in the code block tag string:

Reviewer: Hmm... maybe "might" in that case? "Should" doesn't feel sufficiently future-tense-y to me

> We should allow specifying any of these using the `text` language in the code block tag string:

Author: But I guess this one we definitely will need, so sure, I'll use "will" :)

````markdown
```text path=/third-party/foo/py.typed
partial
```
````

> We may want to also support testing Jupyter notebooks as embedded files using the `json` language.

Reviewer 2: Or, maybe some kind of start and end indicator that signals that any code blocks between them should be in a single Jupyter Notebook file.

> We may want to also support testing Jupyter notebooks as embedded files using the `json` language.

Author: I guess it depends what we need to test about notebooks. If all we care about is the Python contents of cells, then something like @dhruvmanila suggests makes sense. If we want to write tests in this format that actually test other characteristics of the JSON structure of a notebook, then I think we should embed the JSON directly. It doesn't help readability to have key aspects of the test hidden in a different file.

> We may want to also support testing Jupyter notebooks as embedded files using the `json` language.

Reviewer 2: > If we want to write tests in this format that actually test other characteristics of the JSON structure of a notebook

I'm not sure if Red Knot should be testing the notebook JSON structure, I think it should be done as an integration tests which currently there are :)

But yeah, what you said makes sense and we can think about it when it's required.

> We may want to also support testing Jupyter notebooks as embedded files using the `json` language.

Author: Updated the text here to not mention using `json` language and just say exact syntax is TBD.

Of course red-knot is only run directly on `py` and `pyi` files, and assertion comments are only
possible in these files.

(A fenced code block with no language should always be an error.)

### Configuration

We should be able to specify non-default red-knot configurations to use in tests, by including a

> TOML fenced code block:

Author: Same comment as above; I was preferring to avoid definitive "will" language.

> TOML fenced code block:

Author: Ok, looking at it again, I think this is another case where we are pretty committed to doing this.

> TOML fenced code block:

Author: I don't think "TOML-fenced" is correct here. The code block is fenced by triple back ticks, it is not fenced by TOML. It is a "fenced code block" whose language is TOML.

> TOML fenced code block:

Reviewer: Ah I see! In that case I think maybe the adjective order is incorrect/confusing -- probably "fenced TOML code block"?

````markdown

> ```toml
> [tool.knot]
> warn-on-any = true
> 

Reviewer 2: ````suggestion
```toml
[tool.knot]
warn-on-any = true
```

````

```py
from typing import Any

def f(x: Any):  # error: [use-of-any]
    pass
```
````

It should be possible to include a TOML fenced code block in a single test (as shown), or in a
grouping section, in which case it applies to all nested tests within that grouping section. Configs
at multiple level are allowed and merged, with the most-nested (closest to the test) taking
precedence.

### Configuring search paths and kinds

The red-knot TOML config format hasn't been designed yet, and we may want to implement support in
the test framework for configuring search paths before it is designed. If so, we can define some
configuration options for now under the `[tool.knot.tests]` namespace. In the future, perhaps some
of these can be replaced by real red-knot config options, maybe some of them will be kept long term
as test-specific options.

> We should be able to configure the default workspace root to something other than `/src/` using a
> `workspace-root` config option.
> 
> We should be able to add a third-party root using the `third-party-root` config option.
> 
> We may want to add additional config options for setting additional search path kinds.

Reviewer: ```suggestion
Other future planned changes to configuration include:

- We should be able to configure the default workspace root to something other than `/src/` using a
`workspace-root` config option.

- We should be able to add a third-party root using the `third-party-root` config option.

- We may want to add additional config options for setting additional search path kinds.
```

Paths for `workspace-root` and `third-party-root` must be absolute.

Relative embedded-file paths are relative to the workspace root, even if it is explicitly set to a
non-default value using the `workspace-root` config.

> ### Specifying custom typeshed

Reviewer: ```suggestion
### Specifying a custom typeshed
```

Some tests will need to override the default typeshed with custom files. The `[tool.knot.tests]`
config option `typeshed-root` should be usable for this:

````markdown
```toml
[tool.knot.tests]
typeshed-root = "/typeshed"
```

This file is importable as part of our custom typeshed, because it is within `/typeshed`, which we
configured above as our custom typeshed root:

```py path=/typeshed/stdlib/builtins.pyi

> I_AM_THE_ONLY_BUILTIN = 1

Reviewer: scary

```

This file is written to `/src/test.py`, because the default workspace root is `/src/ and the default
file path is `test.py`:

> ```py
> reveal_type(I_AM_THE_ONLY_BUILTIN)  # revealed: Literal[1]
> 

Reviewer 2: ````suggestion
```py
reveal_type(I_AM_THE_ONLY_BUILTIN)  # revealed: Literal[1]
```

````

````

A fenced code block with language `text` can be used to provide a `stdlib/VERSIONS` file in the
custom typeshed root. If no such file is created explicitly, one should be created implicitly
including entries enabling all specified `<typeshed-root>/stdlib` files for all supported Python
versions.

### I/O errors

We could use an `error=` config option in the tag string to make a file cause an I/O error on read.

### Asserting on full diagnostic output

The inline comment diagnostic assertions are useful for making quick, readable assertions about
diagnostics in a particular location. But sometimes we will want to assert on the full diagnostic

> output of a test. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics

Reviewer: ```suggestion
output of a tested Python snippet. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics
```

Author: I'll probably use "embedded Python file" here for consistency

in a file, without impacting the contents of that file by changing a comment in it. In these cases,
a Python fenced code block in a test could be followed by a fenced code block with language

> `output`; this contains the full diagnostic output for the preceding test file:

Reviewer: ```suggestion
`output`; this would contain the full diagnostic output for the preceding test file:
```

````markdown
# full output

```py
x = 1
reveal_type(x)
```

> This is just an example, not a proposal that red-knot would every actually output diagnostics in

Reviewer: ```suggestion
This is just an example, not a proposal that red-knot would ever actually output diagnostics in
```

precisely this format:

```output
test.py, line 1, col 1: revealed type is 'Literal[1]'
```
````

We will want to build tooling to automatically capture and update these “full diagnostic output”

> blocks, when tests are run in an update-output mode (probably specified by env var.)

Reviewer: ```suggestion
blocks, when tests are run in an update-output mode (probably specified by an environment variable.)
```

> By default, an `output` block specifies diagnostic output for the file `<workspace-root>/test.py`.

Reviewer: ```suggestion
By default, an `output` block will specify diagnostic output for the file `<workspace-root>/test.py`.
```

test for the same file path.

### Incremental tests

Some tests should validate incremental checking, by initially creating some files, checking them,
and then modifying/adding/deleting files and checking again.

We should add the capability to create an incremental test by using the `stage=` option on some
fenced code blocks in the test:

````markdown
# Incremental

## modify a file

Initial version of `test.py` and `b.py`:

```py
from b import x
reveal_type(x)
```

```py path=b.py
x = 1
```

Initial expected output for `test.py`:

```output
/src/test.py, line 1, col 1: revealed type is 'Literal[1]'
```

Now in our first incremental stage, modify the contents of `b.py`:

```py path=b.py stage=1
# b.py
x = 2
```

And this is our updated expected output for `test.py` at stage 1:

```output stage=1
/src/test.py, line 1, col 1: revealed type is 'Literal[2]'
```

(One reason to use full-diagnostic-output blocks in this test is that updating
inline-comment diagnostic assertions for `test.py` would require specifying new
contents for `test.py` in stage 1, which we don't want to do in this test.)
````

> Any number of stages can be provided in an incremental test. If a stage re-specifies a filename that

Reviewer: ```suggestion
It will be possible to provide any number of stages in an incremental test. If a stage re-specifies a filename that
```

> was specified in a previous stage (or the initial state), that file is modified. A new filename

Reviewer: ```suggestion
was specified in a previous stage (or the initial stage), that file is modified. A new filename
```

> appearing for the first time in a new stage will create a new file. To delete a previously-created

Reviewer: ```suggestion
appearing for the first time in a new stage will create a new file. To delete a previously created
```

file, specify that file with the tag `delete` in its tag string (in this case, it is an error to

> provide non-empty contents.) Any previously-created files that are not re-specified in a later stage

Reviewer: Closing parens should come before the period unless it's a full sentence inside the parens:

```suggestion
provide non-empty contents). Any previously-created files that are not re-specified in a later stage
```

```py
from typing import reveal_type

reveal_type("foo")  # revealed: Literal["foo"]
```

For convenience, type checkers also pretend that `reveal_type` is a built-in, so that this import is
not required. Using `reveal_type` without importing it issues a diagnostic warning that it was used
without importing it, in addition to the diagnostic revealing the type of the expression.

The `# revealed:` assertion must always match a revealed-type diagnostic, and will also match the
undefined-reveal diagnostic, if present, so it's safe to use `reveal_type` in tests either with or
without importing it. (Style preference is to not import it in tests, unless specifically testing
something about the behavior of importing it.)

#### error

A comment beginning with `# error:` is an assertion that a type checker diagnostic will
be emitted, with text span starting on that line. If the comment is simply `# error:`, this will
match any diagnostic. The matching can be narrowed in three ways:

- `# error: [invalid-assignment]` requires that the matched diagnostic have the rule code
    `invalid-assignment`. (The square brackets are required.)

> - `# error: "Some text"` requires that the matched diagnostic's full message contain the text `Some text`. (The double quotes are required in the assertion comment; they are not part of the matched
>     text.)

Reviewer: nit: some very long lines here!

```suggestion
- `# error: "Some text"` requires that the matched diagnostic's full message contain the text
    `Some text`. (The double quotes are required in the assertion comment; they are not 
    part of the matched text.)
```

Author: Oops, thanks.

- `# error: 8 [rule-code]` or `# error: 8 "Some text"` additionally requires that the matched
    diagnostic's text span begins on column 8 (one-indexed) of this line.

Assertions must contain either a rule code or a contains-text, or both, and may optionally also
include a column. They must come in order: first column, if present; then rule code, if present;

> then contains-text, if present. For example, an assertion using all three would look like `# error: 8 [invalid-assignment] "Some text"`.

Reviewer: ```suggestion
then contains-text, if present. For example, an assertion using all three would look like
`# error: 8 [invalid-assignment] "Some text"`.
```

from b import C
reveal_type(C)  # revealed: Literal[C]
```

```py path=b.py
class C: pass
```
````

Relative file names are always relative to the "workspace root", which is also an import root (that
is, the equivalent of a runtime entry on `sys.path`).

The default workspace root is `/src/`. Currently it is not possible to customize this in a test, but
this is a feature we will want to add in the future.

So the above test creates two files, `/src/test.py` and `/src/b.py`, and sets the workspace root to
`/src/`, allowing `test.py` to import from `b.py` using the module name `b`.

## Multi-test suites

A single test suite (Markdown file) can contain multiple tests, by demarcating them using Markdown
header lines:

````markdown
# Same-file invalid assignment

```py
x: int = "foo"  # error: [invalid-assignment]
```

# Cross-file invalid assignment

```py
from b import y
x: int = y  # error: [invalid-assignment]
```

```py path=b.py
y = "foo"
```
````

This test suite contains two tests, one named "Same-file invalid assignment" and the other named
"Cross-file invalid assignment". The first test involves only a single embedded file, and the second
test involves two embedded files.

> The tests are run independently, in independent in-memory file systems and with new red-knot [Salsa](https://github.com/salsa-rs/salsa)
> databases. This means that each is a from-scratch run of the type checker, with no data persisting from any
> previous test.

Reviewer: ```suggestion
The tests are run independently, in independent in-memory file systems and with new red-knot
[Salsa](https://github.com/salsa-rs/salsa) databases. This means that each is a
from-scratch run of the type checker, with no data persisting from any previous test.
```

(3  # error: 2 [unsupported-operands] <<<
  +
 "foo")  # error: 6 >>>
```

The column assertion `6` on the ending line should be optional.

In cases of overlapping such assertions, resolve ambiguity using more angle brackets: `<<<<` begins
an assertion ended by `>>>>`, etc.

### Non-Python files

Some tests may need to specify non-Python embedded files: typeshed `stdlib/VERSIONS`, `pth` files,
`py.typed` files, `pyvenv.cfg` files...

We will allow specifying any of these using the `text` language in the code block tag string:

````markdown
```text path=/third-party/foo/py.typed
partial
```
````

We may want to also support testing Jupyter notebooks as embedded files; exact syntax for this is
yet to be determined.

Of course, red-knot is only run directly on `py` and `pyi` files, and assertion comments are only
possible in these files.

A fenced code block with no language will always be an error.

### Configuration

We will add the ability to specify non-default red-knot configurations to use in tests, by including
a TOML code block:

````markdown
```toml
[tool.knot]
warn-on-any = true
```

```py
from typing import Any

def f(x: Any):  # error: [use-of-any]
    pass
```
````

> It should be possible to include a TOML-fenced code block in a single test (as shown), or in a

Reviewer: ```suggestion
It should be possible to include a TOML code block in a single test (as shown), or in a
```
